doc/dev/radosgw/bucket_index.rst

```diff
@@ -0,0 +1,69 @@
+
+Buckets in RGW store their list of objects in a bucket index. Each index entry stores just enough metadata (size, etag, mtime, etc.) to serve API requests to list objects. These APIs are `ListObjectsV2`_ and `ListObjectVersions`_ in S3, and `GET Container`_ in Swift.
+
+.. _Consistency Guarantee:
+
+---------------------
+Consistency Guarantee
+---------------------
+
+RGW guarantees read-after-write consistency on object operations. This means that once a client receives a successful response to a write request, then the effects of that write must be visible to subsequent read requests.
+
+For example: if an S3 client sends a PutObject request to overwrite an existing object, followed by a GetObject request to read it back, RGW must not return the previous object's contents. It must either respond with the new object's contents, or with the result of a later object write or delete.
+
+This consistency guarantee applies to all object write requests (PutObject, DeleteObject, PutObjectAcl, etc) and all object read requests (HeadObject, GetObject, ListObjectsV2, etc).
+
+------------------
+Rados Object Model
+------------------
+
+S3/Swift objects, or 'API objects', are stored as rados objects in the rgw.buckets.data pool. Each API object is comprised of a head object and zero or more tail objects. Bucket index objects are stored in the rgw.buckets.index pool.
+
+When writing an object, its head object is written last. This acts as an atomic 'commit' to make it visible to read requests.
+
+-----------------
+Index Transaction
+-----------------
+
+To keep the bucket index consistent, all object writes or deletes must also update the index accordingly. Because the head objects are stored in different rados objects than the bucket indices, we can't update both atomically with a single rados operation. In order to satisfy the `Consistency Guarantee`_ for listing operations, we have to coordinate these two object writes using a three-step bucket index transaction:
+
+#. Prepare a transaction on its bucket index object.
+#. Write or delete the head object.
+#. Commit the transaction on the bucket index object (or cancel the transaction if step 2 fails).
+
+Object writes and deletes may race with each other, so a given object may have more than one prepared transaction at a time. RGW considers an object entry to be 'pending' if there are any outstanding transactions, or 'completed' otherwise.
+
+-------
+Listing
+-------
+
+When listing objects, RGW will read all entries (pending and completed) from the bucket index. For any pending entries, it must check whether the head object exists before including that entry in the final listing.
+
+If an RGW crashes in the middle of an `Index Transaction`_, an index entry may get stuck in this 'pending' state. When bucket listing encounters these pending entries, it also sends information from the head object back to the bucket index so it can update the entry and resolve its stale transactions.
+
```

Comment:
(recommend mentioning that this repair step is the so-called "dir-suggest" mechanism, to help developers)